Merge the demo branch into master (#319)
XanaduAI/strawberryfields: .github/CHANGELOG.md

# Release 0.13.0-dev (development release)

<h3>New features since last release</h3>

* Adds initial support for the Xanadu cloud platform.
  [(#101)](https://github.com/XanaduAI/strawberryfields/pull/101)
  [(#148)](https://github.com/XanaduAI/strawberryfields/pull/148)
  [(#294)](https://github.com/XanaduAI/strawberryfields/pull/294)

  Jobs can now be submitted to the Xanadu cloud platform to be run
  on supported hardware using the new `RemoteEngine`:

  ```python
  from strawberryfields import RemoteEngine
  from strawberryfields import ops
  from strawberryfields.utils import random_interferometer

>   eng = RemoteEngine("chip2", token="AUTHENTICATION_TOKEN")

Author: ```suggestion
  # replace AUTHENTICATION_TOKEN with your Xanadu cloud token
  eng = RemoteEngine("chip2", token="AUTHENTICATION_TOKEN")
```

  prog = sf.Program(8)

  U = random_interferometer(4)

  with prog.context as q:
      ops.S2gate(1.0) | (q[0], q[4])
      ops.S2gate(1.0) | (q[1], q[5])
      ops.S2gate(1.0) | (q[2], q[6])
      ops.S2gate(1.0) | (q[3], q[7])

      ops.Interferometer(U) | q[:4]
      ops.Interferometer(U) | q[4:]

  result = eng.run(prog)
  ```

  For more details, see the

>   [Xanadu cloud platform](https://strawberryfields.readthedocs.io/en/latest/introduction/remote.html)

Author: ```suggestion
  [Xanadu cloud platform](https://strawberryfields.readthedocs.io/en/stable/introduction/remote.html)
```

# Release 0.13.0-dev (development release)

<h3>New features since last release</h3>

> * Adds initial support for the Xanadu cloud platform.

Reviewer: ```suggestion
* Adds initial support for the Xanadu's photonic quantum hardware.
```

  [(#101)](https://github.com/XanaduAI/strawberryfields/pull/101)
  [(#148)](https://github.com/XanaduAI/strawberryfields/pull/148)
  [(#294)](https://github.com/XanaduAI/strawberryfields/pull/294)

  Jobs can now be submitted to the Xanadu cloud platform to be run
  on supported hardware using the new `RemoteEngine`:

  ```python
  from strawberryfields import RemoteEngine
  from strawberryfields import ops
  from strawberryfields.utils import random_interferometer

>   # replace AUTHENTICATION_TOKEN with your Xanadu cloud token

Reviewer: :pray:

>   # replace AUTHENTICATION_TOKEN with your Xanadu cloud token

Reviewer: ```suggestion
  # replace AUTHENTICATION_TOKEN with your Xanadu cloud access token
```

  eng = RemoteEngine("chip2", token="AUTHENTICATION_TOKEN")
  prog = sf.Program(8)

  U = random_interferometer(4)

  with prog.context as q:
      ops.S2gate(1.0) | (q[0], q[4])
      ops.S2gate(1.0) | (q[1], q[5])
      ops.S2gate(1.0) | (q[2], q[6])
      ops.S2gate(1.0) | (q[3], q[7])

      ops.Interferometer(U) | q[:4]
      ops.Interferometer(U) | q[4:]

  result = eng.run(prog)
  ```

  For more details, see the
  [Xanadu cloud platform](https://strawberryfields.readthedocs.io/en/stable/introduction/remote.html)
  tutorial.

* Significantly speeds up the Fock backend of Strawberry Fields,
  through a variety of changes:

>   - The Fock backend now uses The Walrus high performance implementations of

Reviewer: Already merged into master?

Author: Yep, the was quite a large merge conflict in the changelog here, so it looks like a rewrite. (and it is)

    the displacement, squeezing, two-mode squeezing, and beamsplitter operations.
    [(#287)](https://github.com/XanaduAI/strawberryfields/pull/287)
    [(#289)](https://github.com/XanaduAI/strawberryfields/pull/289)

  - Custom tensor contractions which make use of symmetry relations for the beamsplitter
    and the two-mode squeeze gate have been added, as well as more efficient contractions
    for diagonal operations in the Fock basis.
    [(#292)](https://github.com/XanaduAI/strawberryfields/pull/292)

<br>

* New `sf` command line program for configuring Strawberry Fields for access

>   with the Xanadu cloud platform, as well as submitting and executing jobs from the

Reviewer: ```suggestion
  to the Xanadu cloud platform, as well as submitting and executing jobs from the
```

  command line.
  [(#146)](https://github.com/XanaduAI/strawberryfields/pull/146)
  [(#312)](https://github.com/XanaduAI/strawberryfields/pull/312)

  The new Strawberry Fields command line program `sf` provides several utilities
  including:

  * `sf configure [--token] [--local]`: configure the connection to the cloud platform

  * `sf run input [--output FILE]`: submit and execute quantum programs from the command line

>   * `sf --ping`: verify your connection to the Xanadu cloud platform.

Reviewer: ```suggestion
  * `sf --ping`: verify your connection to the Xanadu cloud platform
```

* New configuration functions to load configuration from
  keyword arguments, environment variables, and configuration files.
  [(#298)](https://github.com/XanaduAI/strawberryfields/pull/298)
  [(#306)](https://github.com/XanaduAI/strawberryfields/pull/306)

  This includes the ability to automatically store Xanadu cloud platform
  credentials in a configuration file using the new function

  ```python
  sf.store_account("AUTHENTICATION_TOKEN")
  ```

  as well as from the command line,

  ```bash
  $ sf configure --token AUTHENTICATION_TOKEN
  ```

  Configuration files can be saved globally, or locally on a per-project basis.
  For more details, see the
  [configuration documentation](https://strawberryfields.readthedocs.io/en/stable/introduction/configuration.html)

* Adds the `x_quad_values` and `p_quad_values` methods to the `state` class.
  This allows calculation of x and p quadrature
  probability distributions by integrating across the Wigner function.
  [(#270)](https://github.com/XanaduAI/strawberryfields/pull/270)

* Adds support in the applications layer for node-weighted graphs. Users can sample from graphs

>   with node weights using the WAW encoding and input node weights into search algorithms in the

Reviewer: Users might not know what the WAW encoding is

Author: I suggest we address this is a new smaller PR after `demo` is merged, as it is a pre-existing changelog entry.

Author: @trbromley

Reviewer 2: :thinking: Currently the user could look at the linked PRs to find out
To include it here, are we able to have math equations in markdown?
(but yes I agree this could be done outside of this PR)

Reviewer: Easiest thing is to not give any specifics that we have to define. Maybe "sample from graphs with node weights using special-purpose encodings"

  `clique` and `subgraph` modules.
  [(#295)](https://github.com/XanaduAI/strawberryfields/pull/295)
  [(#296)](https://github.com/XanaduAI/strawberryfields/pull/296)
  [(#297)](https://github.com/XanaduAI/strawberryfields/pull/297)

<h3>Improvements</h3>

* Moved Fock backend apply-gate functions to `Circuit` class, and removed
  `apply_gate_einsum` and `Circuits._apply_gate`, since they were no longer used.
  [(#293)](https://github.com/XanaduAI/strawberryfields/pull/293/)

* Results returned from all backends now have a unified type and shape.
  In addition, attempting to use batching, post-selection and feed-foward together
  with multiple shots now raises an error.
  [(#300)](https://github.com/XanaduAI/strawberryfields/pull/300)

* Modified the rectangular decomposition to ensure that identity-like
  unitaries are implemented with no swaps.

>   [#311](https://github.com/XanaduAI/strawberryfields/pull/311)

Reviewer 3: I assume this should also be enclosed in parentheses.

```suggestion
  [(#311)](https://github.com/XanaduAI/strawberryfields/pull/311)
```

<h3>Bug fixes</h3>

> * Symbolic Operation parameters are now compatible with TensorFlow 2.0 objects.

Reviewer: But why would a user want to use a tensorflow 2.0 object? :thinking:

  [(#282)](https://github.com/XanaduAI/strawberryfields/pull/282)

* Added `sympy>=1.5` to the list of dependencies.
  Removed the `sympy.functions.atan2` workaround now that SymPy has been fixed.
  [(#280)](https://github.com/XanaduAI/strawberryfields/pull/280)

> * Removed two unnecessary else statements that pylint complained about.

Reviewer: ```suggestion
* Removed two unnecessary else statements.
```
imo, not even worth putting in changelog

Author: agree

  [(#290)](https://github.com/XanaduAI/strawberryfields/pull/290)

* Fixed a bug in the `MZgate`, where the internal and external phases were
  in the wrong order in both the docstring and the argument list. The new
  signature is `MZgate(phase_in, phase_ex)`, matching the existing `rectangular_symmetric`
  decomposition.
  [(#301)](https://github.com/XanaduAI/strawberryfields/pull/301)

<h3>Contributors</h3>

This release contains contributions from (in alphabetical order):

> Ville Bergholm, Tom Bromley, Jack Ceroni, Theodor Isacsson, Josh Izaac, Shreya P Kumar,

Reviewer: ```suggestion
Ville Bergholm, Tom Bromley, Jack Ceroni, Theodor Isacsson, Josh Izaac, Nathan Killoran, Shreya P Kumar,
```
Don't forget code reviewers!
